examples/bfgs.dx

```diff
@@ -0,0 +1,208 @@
+    a_last = get a_ref_last
+    f_i = f a_i
+    if i >= maxiter || f_i > (f_0 + c1 * a_i * g_0) || (f_i >= f a_last && i > 0)
+      then Done (zoom f a_last a_i c1 c2)
+      else
+        g_i = grad f a_i
+        if abs g_i <= (-c2 * g_0)
+          then Done a_i
+          else
+            if g_i >= 0.
+              then Done (zoom f a_i a_last c1 c2)
+              else 
+                a_ref_last := a_i
+                a_ref := 0.5 * (a_i + a_max)
+                Continue
+
+def backtracking_line_search(
+  maxiter:Nat,
+  f: (Float)->Float
+  ) -> Float =
+  -- Algorithm 3.1 in Nocedal and Wright (2006).
+  a_init = 1. 
+  f_0 = f 0.
+  g_0 = grad f 0.
+  rho = 0.5
+  c = 0.0001
+
+  a_ref <- with_state a_init
+  iter \i.
+    a_i = get a_ref
+    if f a_i <= (f_0 + c * a_i * g_0) || i >= maxiter
+      then Done a_i
+      else
+        a_ref := a_i * rho
+        Continue
+
+struct BFGSresults(n|Ix) = 
+  fval : Float
+  x_opt: (n=>Float)
+  error: Float 
+  num_iter: Nat
+  
+def bfgs_minimize(
+  f: (n=>Float)->Float,                 --Objective function.
+  x0: n=>Float,                       --Initial point.
+  H0: n=>n=>Float,                    --Initial inverse Hessian approximation.
+  linesearch: ((Float)->Float)->Float,  --Line search that returns a step size.
+  tol: Float,                         --Convergence tolerance (of the gradient L2 norm).
+  maxiter: Nat                        --Maximum number of BFGS iterations.
+  ) -> BFGSresults n given (n|Ix) =             
+  -- Algorithm 6.1 in Nocedal and Wright (2006).
+
+  xref <- with_state x0 
+  Href <- with_state H0
+  gref <- with_state (grad f x0)
+
+  iter \i.
+    x = get xref
+    g = get gref
+    grad_norm = sqrt $ sum $ for i. g[i] * g[i]
+
+    if grad_norm < tol || i >= maxiter
+      then Done BFGSresults(f x, x, grad_norm, i)
+      else
+        H = get Href
+        search_direction = -H**.g
+        f_line = \s:Float. f (x + s .* search_direction)
+        step_size = linesearch f_line 
+        x_diff = step_size .* search_direction
+        x_next = x + x_diff
+        g_next = grad f x_next
+        grad_diff = g_next - g
+
+        -- Update the inverse Hessian approximation.
+        rho_inv = vdot grad_diff x_diff
+        if not (rho_inv == 0.)
+            then
+                rho = 1. / rho_inv
+                y = (eye - rho .* outer_product x_diff grad_diff)
+                Href := y ** H ** (transpose y) + rho .* outer_product x_diff x_diff
+        
+        xref := x_next
+        gref := g_next
+        Continue
+
+
+def rosenbrock(coord:(Fin 2)=>Float) -> Float =
+  --A standard optimization test case; the optimum is (1, 1).
+  x = coord[0@_]
+  y = coord[1@_]
+  pow (1 - x) 2 + 100 * pow (y - x * x) 2
+
+-- %bench "rosenbrock"
+bfgs_minimize rosenbrock [10., 10.] eye (\f. backtracking_line_search 15 f) 0.001 100
+
+def multiclass_logistic_loss(xs: n=>d=>Float, ys: n=>Nat, w: (d, m)=>Float) -> Float given (n|Ix, d|Ix, m|Ix) =
```

Comment:
What do you think of declaring `ys: n=>m` instead of `ys: n=>Nat`?  It avoids the potentially failing conversion `ys[i]@_` three lines down, and generally ensures that the weight matrix exactly covers the extant labels.

Then in `multiclass_logreg` you don't need to take `num_classes` as a separate argument, since it will be apparent from the type of `ys`.  And then in `multiclass_logreg_int`, you can specialize the element type of `ys` to be `Fin num_classes`.

Reply:
Nice, I had a feeling there was a way to clean it up along those lines but I didn’t quite see it.